kinesis/kinesis.go

```diff
@@ -185,6 +189,32 @@ func (outputPlugin *OutputPlugin) Flush() error {
 	return outputPlugin.sendCurrentBatch()
 }
 
+func replaceDots(obj map[interface{}]interface{}) map[interface{}]interface{} {
+	newObj := make(map[interface{}]interface{})
+
+	for k, v := range obj {
+		switch kt := k.(type) {
+		case string:
+			k = strings.ReplaceAll(kt, ".", "_")
+		}
+
+		switch vt := v.(type) {
+		case map[interface{}]interface{}:
+			v = replaceDots(vt)
+			// Ensure duplicated keys are merged rather than replaced
+			if _, ok := newObj[k]; ok {
```

Comment:
checking for duplicated keys needs to happen in the first call to `replaceDots`, not just in the recursive case

Reply:
Interestingly, the C code for the ES version of this option is very simple and does not check for duplicates:

https://github.com/fluent/fluent-bit/blob/master/plugins/out_es/es.c#L90

Reply:
Checking for duplicated keys is safer; I like having that check. However, I think it should be able to be done in place without allocating a new map.

I believe this should be safe in Go: https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop